server/coordinator/procedure/transfer_leader.go

```diff
@@ -0,0 +1,179 @@
+
+	transferLeaderRequest := &TransferLeaderCallbackRequest{
+		cluster:   p.cluster,
+		ctx:       ctx,
+		newLeader: p.newLeader,
+		oldLeader: p.oldLeader,
+		dispatch:  p.dispatch,
+	}
+
+	if err := p.fsm.Event(eventTransferLeaderPrepare, transferLeaderRequest); err != nil {
+		err := p.fsm.Event(eventTransferLeaderFailed, transferLeaderRequest)
+		p.UpdateStateWithLock(StateFailed)
+		return errors.WithMessage(err, "coordinator transferLeaderShard start")
+	}
+
+	if err := p.fsm.Event(eventTransferLeaderSuccess, transferLeaderRequest); err != nil {
+		return errors.WithMessage(err, "coordinator transferLeaderShard start")
+	}
+
+	p.UpdateStateWithLock(StateFinished)
+	return nil
+}
+
+func (p *TransferLeaderProcedure) Cancel(_ context.Context) error {
+	p.UpdateStateWithLock(StateCancelled)
+	return nil
+}
+
+func (p *TransferLeaderProcedure) State() State {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+	return p.state
+}
+
+func transferLeaderPrepareCallback(event *fsm.Event) {
+	request := event.Args[0].(*TransferLeaderCallbackRequest)
+	ctx := request.ctx
+
+	closeShardRequest := &eventdispatch.CloseShardRequest{
+		ShardID: request.oldLeader.Id,
+	}
+	if err := request.dispatch.CloseShard(ctx, request.oldLeader.Node, closeShardRequest); err != nil {
+		CancelEventWithLog(event, err, LogLevelError, "close shard failed", zap.Uint32("shardId", request.oldLeader.Id))
+		return
+	}
+
+	openShardRequest := &eventdispatch.OpenShardRequest{
+		Shard: &cluster.ShardInfo{ShardID: request.newLeader.Id, ShardRole: clusterpb.ShardRole_LEADER},
+	}
+	if err := request.dispatch.OpenShard(ctx, request.newLeader.Node, openShardRequest); err != nil {
+		CancelEventWithLog(event, err, LogLevelError, "open shard failed", zap.Uint32("shardId", request.newLeader.Id))
+		return
+	}
+}
+
+func transferLeaderFailedCallback(_ *fsm.Event) {
+	// TODO: Use RollbackProcedure to rollback transfer failed
+}
+
+func transferLeaderSuccessCallback(event *fsm.Event) {
+	request := event.Args[0].(*TransferLeaderCallbackRequest)
+	c := request.cluster
+	ctx := request.ctx
+
+	// Update cluster topology
+	shardView, err := c.GetClusterShardView()
+	if err != nil {
+		CancelEventWithLog(event, err, LogLevelError, "get cluster shard view failed")
+		return
+	}
+	oldLeaderIndex := -1
+	for i := 0; i < len(shardView); i++ {
+		shardID := shardView[i].Id
+		if shardID == request.oldLeader.Id {
+			oldLeaderIndex = i
+		}
+	}
+	if oldLeaderIndex == -1 {
+		event.Cancel(errors.WithMessage(cluster.ErrShardNotFound, fmt.Sprintf("shard not found,shardID = %d", request.oldLeader.Id)))
+		return
+	}
+	shardView = append(shardView[:oldLeaderIndex], shardView[oldLeaderIndex+1:]...)
+	shardView = append(shardView, request.newLeader)
+
+	if err := c.UpdateClusterTopology(ctx, c.GetClusterState(), shardView); err != nil {
+		CancelEventWithLog(event, err, LogLevelError, "update shard topology failed")
+		return
+	}
+}
+
+func (p *TransferLeaderProcedure) UpdateStateWithLock(state State) {
```

Comment:
Maybe you should check all the public methods/structs to ensure the visibility as it should be.